File: src/classifier/dsl.ts
import { ageInHours, folderName, type Rule } from './rule-classifier.js';
import type { Decision, MailMessage } from '../types.js';

/**
 * A tiny declarative condition language for rules, so the rule logic itself can live in
 * rules.yaml instead of TypeScript. A condition is either a leaf test or a combinator:
 *
 *   leaf:        { field, op, value }   or   { field, op, ref }
 *   combinator:  { all: [...] } | { any: [...] } | { not: <condition> }
 *
 * `value` is an inline literal; `ref` is a dotted path resolved within rules.yaml's data
 * (e.g. `lists.markReadDomains`, `patterns.spamSubject`, `thresholds.monthly`).
 *
 * Conditions are compiled once into a predicate (regexes/refs resolved up front), so
 * per-message evaluation is cheap. Compilation throws on any malformed rule.
 */

/** Message fields exposed to the DSL. */
const STRING_FIELDS = {
  subject: (m: MailMessage) => m.subject,
  fromAddress: (m: MailMessage) => m.fromAddress,
  fromName: (m: MailMessage) => m.fromName,
  /** subject + body, for rules that should look at content too. */
  content: (m: MailMessage) => `${m.subject} ${m.bodyPreview}`,
  /** The LEAF folder name, so `GitHub` matches `Archive/GitHub`. */
  folder: (m: MailMessage) => folderName(m),
} as const;

const NUMBER_FIELDS = {
  /** Whole-message age in hours (thresholds are stored in hours too). */
  age: (m: MailMessage) => ageInHours(m),
} as const;

export type LeafCondition = {
  field: string;
  op: string;
  value?: unknown;
  ref?: string;
};
export type Condition =
  | LeafCondition
  | { all: Condition[] }
  | { any: Condition[] }
  | { not: Condition };

export type ResolveRef = (path: string) => unknown;

type Predicate = (message: MailMessage) => boolean;

function stringField(field: string, op: string): (m: MailMessage) => string {
  const get = (STRING_FIELDS as Record<string, (m: MailMessage) => string>)[field];
  if (!get) {
    throw new Error(`rules.yaml: op "${op}" needs a text field, got "${field}".`);
  }
  return get;
}

function numberField(field: string, op: string): (m: MailMessage) => number {
  const get = (NUMBER_FIELDS as Record<string, (m: MailMessage) => number>)[field];
  if (!get) {
    throw new Error(`rules.yaml: op "${op}" needs a numeric field, got "${field}".`);
  }
  return get;
}

/** The literal (`value`) or resolved (`ref`) operand of a leaf condition. */
function operand(leaf: LeafCondition, resolveRef: ResolveRef): unknown {
  if (leaf.ref !== undefined) {
    return resolveRef(leaf.ref);
  }
  if (leaf.value === undefined) {
    throw new Error(`rules.yaml: leaf {field:${leaf.field}, op:${leaf.op}} needs a value or ref.`);
  }
  return leaf.value;
}

function asStringList(value: unknown, op: string): string[] {
  if (!Array.isArray(value)) {
    throw new Error(`rules.yaml: op "${op}" needs a list operand.`);
  }
  return value.map((v) => String(v).toLowerCase());
}

/** Compiles a leaf test into a predicate. String ops are case-insensitive; equals is exact. */
function compileLeaf(leaf: LeafCondition, resolveRef: ResolveRef): Predicate {
  const { field, op } = leaf;
  const value = operand(leaf, resolveRef);

  switch (op) {
    case 'equals': {
      const get = stringField(field, op);
      const target = String(value);
      return (m) => get(m) === target;
    }
    case 'includes': {
      const get = stringField(field, op);
      const needle = String(value).toLowerCase();
      return (m) => get(m).toLowerCase().includes(needle);
    }
    case 'endsWith': {
      const get = stringField(field, op);
      const suffix = String(value).toLowerCase();
      return (m) => get(m).toLowerCase().endsWith(suffix);
    }
    case 'endsWithAny': {
      const get = stringField(field, op);
      const suffixes = asStringList(value, op);
      return (m) => {
        const text = get(m).toLowerCase();
        return suffixes.some((s) => text.endsWith(s));
      };
    }
    case 'regex': {
      const get = stringField(field, op);
      const re = value instanceof RegExp ? value : new RegExp(String(value), 'i');
      return (m) => re.test(get(m));
    }
    case '>': {
      const get = numberField(field, op);
      const n = Number(value);
      return (m) => get(m) > n;
    }
    case '>=': {
      const get = numberField(field, op);
      const n = Number(value);
      return (m) => get(m) >= n;
    }
    default:
      throw new Error(`rules.yaml: unknown op "${op}".`);
  }
}

/** Recursively compiles any condition node into a single predicate. */
export function compileCondition(node: Condition, resolveRef: ResolveRef): Predicate {
  if (node && typeof node === 'object') {
    if ('all' in node) {
      const subs = node.all.map((n) => compileCondition(n, resolveRef));
      return (m) => subs.every((p) => p(m));
    }
    if ('any' in node) {
      const subs = node.any.map((n) => compileCondition(n, resolveRef));
      return (m) => subs.some((p) => p(m));
    }
    if ('not' in node) {
      const sub = compileCondition(node.not, resolveRef);
      return (m) => !sub(m);
    }
    if ('field' in node && 'op' in node) {
      return compileLeaf(node, resolveRef);
    }
  }
  throw new Error(`rules.yaml: malformed condition: ${JSON.stringify(node)}`);
}

const DECISIONS = new Set<Decision>(['keep', 'markRead', 'delete']);

interface RawRule {
  name?: unknown;
  when?: unknown;
  then?: unknown;
}

/** Compiles the declarative rule array from rules.yaml into executable {@link Rule}s. */
export function compileRules(raw: unknown, resolveRef: ResolveRef): Rule[] {
  if (!Array.isArray(raw)) {
    throw new Error('rules.yaml: expected a list of rules.');
  }
  return raw.map((entry, index) => {
    const { name, when, then } = entry as RawRule;
    if (typeof name !== 'string') {
      throw new Error(`rules.yaml: rule #${index} is missing a string "name".`);
    }
    if (!DECISIONS.has(then as Decision)) {
      throw new Error(`rules.yaml: rule "${name}" has invalid "then": ${JSON.stringify(then)}.`);
    }
    if (when === undefined) {
      throw new Error(`rules.yaml: rule "${name}" is missing "when".`);
    }
    return { name, when: compileCondition(when as Condition, resolveRef), then: then as Decision };
  });
}

File: crates/mailward-core/src/rules/dsl.rs
//! The declarative condition language. A condition is either a leaf test or a
//! combinator; conditions compile once into a [`Predicate`] (regexes/refs
//! resolved up front) so per-message evaluation is cheap.
//!
//! ```yaml
//! when:
//!   all:
//!     - { field: folder, op: equals, value: Junk }
//!     - { field: subject, op: regex, ref: patterns.spamSubject }
//! ```

use std::borrow::Cow;

use regex::{Regex, RegexBuilder};
use serde::Deserialize;
use serde_yaml_ng::Value;
use unicode_normalization::UnicodeNormalization;

use super::RuleError;
use super::data::{RefValue, RulesData};
use crate::message::{MailMessage, folder_name};

/// Message fields exposed to the DSL.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Field {
    Subject,
    FromAddress,
    FromName,
    /// subject + body preview, for rules that look at content.
    Content,
    /// the LEAF folder name, so `GitHub` matches `Archive/GitHub`.
    Folder,
    /// whole-message age in hours.
    Age,
}

/// Comparison operators. String ops are case-insensitive except `equals` (exact).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
pub enum Op {
    #[serde(rename = "regex")]
    Regex,
    #[serde(rename = "includes")]
    Includes,
    #[serde(rename = "endsWith")]
    EndsWith,
    #[serde(rename = "endsWithAny")]
    EndsWithAny,
    #[serde(rename = "equals")]
    Equals,
    #[serde(rename = ">")]
    Gt,
    #[serde(rename = ">=")]
    Gte,
}

/// A single leaf test: `{ field, op, value }` or `{ field, op, ref }`.
#[derive(Debug, Clone, Deserialize)]
pub struct LeafCondition {
    pub field: Field,
    pub op: Op,
    #[serde(default)]
    pub value: Option<Value>,
    #[serde(default, rename = "ref")]
    pub reference: Option<String>,
}

/// A condition node: a leaf or a combinator. Deserialized untagged - a map with
/// `all` / `any` / `not` is a combinator, otherwise it's a leaf.
#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum Condition {
    All { all: Vec<Condition> },
    Any { any: Vec<Condition> },
    Not { not: Box<Condition> },
    Leaf(LeafCondition),
}

/// A compiled, ready-to-evaluate condition.
pub type Predicate = Box<dyn Fn(&MailMessage) -> bool + Send + Sync>;

/// Recursively compiles a condition into a single [`Predicate`].
pub fn compile_condition(node: &Condition, data: &RulesData) -> Result<Predicate, RuleError> {
    match node {
        Condition::All { all } => {
            let subs = compile_all(all, data)?;
            Ok(Box::new(move |m| subs.iter().all(|p| p(m))))
        }
        Condition::Any { any } => {
            let subs = compile_all(any, data)?;
            Ok(Box::new(move |m| subs.iter().any(|p| p(m))))
        }
        Condition::Not { not } => {
            let sub = compile_condition(not, data)?;
            Ok(Box::new(move |m| !sub(m)))
        }
        Condition::Leaf(leaf) => compile_leaf(leaf, data),
    }
}

fn compile_all(nodes: &[Condition], data: &RulesData) -> Result<Vec<Predicate>, RuleError> {
    nodes.iter().map(|n| compile_condition(n, data)).collect()
}

fn compile_leaf(leaf: &LeafCondition, data: &RulesData) -> Result<Predicate, RuleError> {
    let field = leaf.field;
    let operand = resolve_operand(leaf, data)?;

    Ok(match leaf.op {
        Op::Equals => {
            ensure_text_field(field, leaf.op)?;
            let target = operand.into_text(leaf.op)?;
            Box::new(move |m| string_value(m, field).as_ref() == target)
        }
        Op::Includes => {
            ensure_text_field(field, leaf.op)?;
            let needle = operand.into_text(leaf.op)?.to_lowercase();
            Box::new(move |m| string_value(m, field).to_lowercase().contains(&needle))
        }
        Op::EndsWith => {
            ensure_text_field(field, leaf.op)?;
            let suffix = operand.into_text(leaf.op)?.to_lowercase();
            Box::new(move |m| string_value(m, field).to_lowercase().ends_with(&suffix))
        }
        Op::EndsWithAny => {
            ensure_text_field(field, leaf.op)?;
            let suffixes: Vec<String> = operand
                .into_list(leaf.op)?
                .into_iter()
                .map(|s| s.to_lowercase())
                .collect();
            Box::new(move |m| {
                let text = string_value(m, field).to_lowercase();
                suffixes.iter().any(|s| text.ends_with(s))
            })
        }
        Op::Regex => {
            ensure_text_field(field, leaf.op)?;
            let regex = operand.into_regex(leaf.op)?;
            Box::new(move |m| regex.is_match(string_value(m, field).as_ref()))
        }
        Op::Gt => {
            ensure_age_field(field, leaf.op)?;
            let threshold = operand.into_num(leaf.op)?;
            Box::new(move |m| number_value(m, field) > threshold)
        }
        Op::Gte => {
            ensure_age_field(field, leaf.op)?;
            let threshold = operand.into_num(leaf.op)?;
            Box::new(move |m| number_value(m, field) >= threshold)
        }
    })
}

/// The literal (`value`) or resolved (`ref`) operand of a leaf, as owned data the
/// compiled predicate can capture.
enum Operand {
    Text(String),
    Num(f64),
    List(Vec<String>),
    Pattern(Regex),
}

impl Operand {
    fn into_text(self, op: Op) -> Result<String, RuleError> {
        match self {
            Operand::Text(s) => Ok(s),
            Operand::Num(n) => Ok(format!("{n}")),
            _ => Err(operand_type(op, "a text")),
        }
    }

    fn into_list(self, op: Op) -> Result<Vec<String>, RuleError> {
        match self {
            Operand::List(v) => Ok(v),
            _ => Err(operand_type(op, "a list")),
        }
    }

    fn into_num(self, op: Op) -> Result<f64, RuleError> {
        match self {
            Operand::Num(n) => Ok(n),
            Operand::Text(s) => s.parse().map_err(|_| operand_type(op, "a numeric")),
            _ => Err(operand_type(op, "a numeric")),
        }
    }

    fn into_regex(self, op: Op) -> Result<Regex, RuleError> {
        match self {
            Operand::Pattern(re) => Ok(re),
            Operand::Text(s) => RegexBuilder::new(&s)
                .case_insensitive(true)
                .build()
                .map_err(|source| RuleError::BadRegex {
                    key: format!("(inline {op:?})"),
                    source,
                }),
            _ => Err(operand_type(op, "a regex")),
        }
    }
}

fn resolve_operand(leaf: &LeafCondition, data: &RulesData) -> Result<Operand, RuleError> {
    if let Some(path) = &leaf.reference {
        return Ok(match data.resolve(path)? {
            RefValue::List(v) => Operand::List(v.to_vec()),
            RefValue::Pattern(re) => Operand::Pattern(re.clone()),
            RefValue::Threshold(n) => Operand::Num(n),
        });
    }
    if let Some(value) = &leaf.value {
        return value_to_operand(value);
    }
    Err(RuleError::MissingOperand {
        field: format!("{:?}", leaf.field),
        op: format!("{:?}", leaf.op),
    })
}

fn value_to_operand(value: &Value) -> Result<Operand, RuleError> {
    Ok(match value {
        Value::String(s) => Operand::Text(s.clone()),
        Value::Bool(b) => Operand::Text(b.to_string()),
        Value::Number(n) => Operand::Num(n.as_f64().unwrap_or_default()),
        Value::Sequence(seq) => Operand::List(seq.iter().filter_map(value_scalar).collect()),
        other => {
            return Err(RuleError::OperandType {
                op: "value".into(),
                expected: format!("a scalar or list (got {other:?})"),
            });
        }
    })
}

fn value_scalar(value: &Value) -> Option<String> {
    match value {
        Value::String(s) => Some(s.clone()),
        Value::Number(n) => Some(n.to_string()),
        Value::Bool(b) => Some(b.to_string()),
        _ => None,
    }
}

/// NFKC-fold human-authored text so Unicode "stylised" letters spammers use to dodge
/// plain-text rules (mathematical bold `𝙉𝙤`, fullwidth `Ｎｏ`, ...) collapse to their ASCII
/// form before matching. ASCII is the common case and returns borrowed (no allocation);
/// accents (`ä`) are preserved, since NFKC only undoes *compatibility* decompositions.
fn fold_for_match(text: &str) -> Cow<'_, str> {
    if text.is_ascii() {
        Cow::Borrowed(text)
    } else {
        Cow::Owned(text.nfkc().collect())
    }
}

/// The text of a field for the current message (case folding happens at the call site).
/// Human-text fields are confusables-folded; addresses and folder names are matched raw.
fn string_value(message: &MailMessage, field: Field) -> Cow<'_, str> {
    match field {
        Field::Subject => fold_for_match(&message.subject),
        Field::FromAddress => Cow::Borrowed(&message.from_address),
        Field::FromName => fold_for_match(&message.from_name),
        Field::Content => {
            let combined = format!("{} {}", message.subject, message.body_preview);
            if combined.is_ascii() {
                Cow::Owned(combined)
            } else {
                Cow::Owned(combined.nfkc().collect())
            }
        }
        Field::Folder => Cow::Borrowed(folder_name(&message.folder)),
        // Guarded by `ensure_text_field`; never reached.
        Field::Age => Cow::Borrowed(""),
    }
}

fn number_value(message: &MailMessage, field: Field) -> f64 {
    match field {
        Field::Age => message.age_hours,
        // Guarded by `ensure_age_field`; never reached.
        _ => 0.0,
    }
}

fn ensure_text_field(field: Field, op: Op) -> Result<(), RuleError> {
    if field == Field::Age {
        return Err(RuleError::WrongField {
            op: format!("{op:?}"),
            expected: "a text".into(),
            field: format!("{field:?}"),
        });
    }
    Ok(())
}

fn ensure_age_field(field: Field, op: Op) -> Result<(), RuleError> {
    if field != Field::Age {
        return Err(RuleError::WrongField {
            op: format!("{op:?}"),
            expected: "a numeric".into(),
            field: format!("{field:?}"),
        });
    }
    Ok(())
}

fn operand_type(op: Op, expected: &str) -> RuleError {
    RuleError::OperandType {
        op: format!("{op:?}"),
        expected: expected.into(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn message(subject: &str) -> MailMessage {
        MailMessage {
            id: "1".into(),
            folder: "Junk".into(),
            subject: subject.into(),
            from_address: String::new(),
            from_name: String::new(),
            body_preview: String::new(),
            age_hours: 0.0,
            is_read: false,
        }
    }

    #[test]
    fn fold_keeps_ascii_borrowed_and_preserves_accents() {
        assert!(matches!(fold_for_match("plain ascii"), Cow::Borrowed(_)));
        // NFKC must NOT strip the German umlauts the rules rely on (bestätigt, gefährdet, ...).
        assert_eq!(fold_for_match("bestätigt").as_ref(), "bestätigt");
    }

    #[test]
    fn fold_collapses_unicode_confusables_to_ascii() {
        // Mathematical sans-serif bold - a classic spam evasion - folds to ASCII.
        assert_eq!(fold_for_match("𝙉𝙤 𝘿𝙚𝙥𝙤𝙨𝙞𝙩").as_ref(), "No Deposit");
        // Fullwidth Latin folds too.
        assert_eq!(fold_for_match("Ｆｒｅｅ Ｓｐｉｎｓ").as_ref(), "Free Spins");
    }

    #[test]
    fn subject_is_folded_so_styled_spam_still_matches() {
        let m = message("Claim 50 𝙁𝙧𝙚𝙚 𝙎𝙥𝙞𝙣𝙨 now");
        assert_eq!(
            string_value(&m, Field::Subject).as_ref(),
            "Claim 50 Free Spins now"
        );
    }
}
